crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -486,32 +486,15 @@ where
                     }
                 }
             } else if let Some(signer) = changed_balances.decoder.get(&signer) {
-                // Otherwise the signer must be decodable so that we can
-                // manually check the signatures
-                let public_keys_index_map =
-                    crate::account::public_keys_index_map(
-                        &self.ctx.pre(),
-                        signer,
-                    )?;
-                let threshold =
-                    crate::account::threshold(&self.ctx.pre(), signer)?
-                        .unwrap_or(1);
-                let max_signatures_per_transaction =
-                    crate::parameters::max_signatures_per_transaction(
-                        &self.ctx.pre(),
-                    )?;
-                let mut gas_meter = self.ctx.gas_meter.borrow_mut();
-                tx_data
-                    .tx
-                    .verify_signatures(
-                        &[tx_data.tx.raw_header_hash()],
-                        public_keys_index_map,
-                        &Some(signer.clone()),
-                        threshold,
-                        max_signatures_per_transaction,
-                        || gas_meter.consume(crate::gas::VERIFY_TX_SIG_GAS),
-                    )
-                    .map_err(native_vp::Error::new)?;
+                // Otherwise the owner's vp must have been triggered
+                if !verifiers.contains(signer) {
```

Comment:
I would almost certainly agree with this logic if `signer`'s VP was being executed on the state change being checked by this MASP VP minus the state changes induced by the inner `Transaction`. My concern here is as follows: imagine Alice constructs an unshielding `Transaction` of 10 BTC to Bertha. Now imagine that Christel intercepts this `Transaction` and embeds it inside a  `Transfer` `Tx` that increases the balance of Christel by 10 BTC (instead of Bertha who is the rightful recipient). In this case, since `Tx` is effectively two transfers (an unshielding `Transaction` of 10 BTC to Bertha followed by a transparent `Transfer` of 10 BTC from Bertha to Christel), the `signers` map would contain Bertha's address. My specific concern here is that Bertha's VP doesn't see this foul play (of rehousing the `Transaction` in a different `Tx`) and always accepts the state change because the net change to Bertha's balance as a result of the entire state change is zero (after all, Bertha is only mentioned in the transparent output of the `Transaction`). To me this approach would only make sense if the verifiers could be made to specifically validate just the `Transfer` of 10 BTC from Bertha to Christel (which would be the result of doing the above "subtraction"), in which case they would certainly demand an authorization signature from Bertha (as was being done before in an admittedly restrictive way). Am I missing something here?

Reply:
So this PR only removes the signature verification and requests the corresponding VP to be triggered. As you are saying though, Bertha's VP would not reject the tx in the attack scenario you described because the balance change is 0. To make this work I've introduced a new masp `Action` (i.e. a temporary write to write log that is not committed to storage) with the address of the verifier whose vp we want to check this tx (in this case Bertha). Again, the 0 balance change would lead to the tx being accepted, so on top of that, I've updated the implicit and user VPs to handle this `Action` with a signature check (but custom vps could add more checks if desired): so the signature check has been simply moved from the masp vp to the user and implicit vps.

For this logic to work we need some cooperation from the tx which must write the temporary action key(s) required (the ones carrying the addresses whose extra signatures have been attached to the tx).

I believe this logic is ok (but please double check again) but I think that there are other flaws in this this implementation of mine:
1. Even though the `Action` carries the address of the signer this does not trigger its vp, since it's not the leading part of the key and we don't trigger VPs anymore if their address appear in non-leading locations of a key (@tzemanovic  please correct me if I'm wrong). In this case we need a way for the tx to trigger the required VPs
2. Checking that the VPs have been triggered is not enough. Someone could craft a tx that does not write the required actions but for which the relevant VPs are still triggered. In this case we could end up in the scenario you envisioned. On top of checking that VPs have been triggered we should also check that the corresponding `Action` has been written to storage to ensure that the VP will run the correct validation

Note that, as I've already answered Tomas before, at the moment the transactions do not support this `Action` logic cause with the previous design of vectorized transfers these scenario could not happen. It will become relevant after the new vectorized strategy has been merged

Reply:
For a single transfer I think we would need do something like this:

1. If a VP of an address who must authorize the tx is not going to be triggered from the changed keys (yes, it is only the leading part) then the tx has to `insert_verifier`
2. The MASP VP should require an `Action` to be written so a tx is not allowed to skip it (i.e. actions must not be empty)

For these to work together the MASP VP should then ensure that when the action is `MaspSigner` then the address in this action is contained in the `verifiers` set.

*However*, with the scenario with multiple transfers @murisi described I think this may not be sufficient. A tx could only attach a single action despite performing multiple actions. We might then need to add more information to the actions to describe the transfers in more detail and the MASP and user/implicit VPs will have to enforce that the storage changes match these actions. In this example, for the unshielding action from Alice we could have e.g. `Action::Unshield { dest, amount }` where `dest == Bertha`` and if `Christel` was to maliciously embed this transfer in another tx, either:

- Only with the original action, an increase in Christel's balance would not be matching the action (the MASP VP has to check the transparent bundle's `vout.address` against the actions to enforce this)
- With the original action and added `Action::Transparent { src, dest, amount } where `src == Bertha`, `dest == Christel` and `amount == 10`, the user/implicit VP must require signature from Bertha (even if the overall balance change of Bertha is 0)
- Without the original action, but with an `Action::Unshield` where `dest == Christel`, the side-effects of the storage changes would not match the actions (the actual unshielding `dest` is Bertha)

Reply:
> So this PR only removes the signature verification and requests the corresponding VP to be triggered.

Ah, I completely misread the approach taken by this PR. My bad.
> but custom vps could add more checks if desired

I see. And at the bare minimum, custom VPs must not forget to handle MASP `Action`.
> In this case we need a way for the tx to trigger the required VPs

Nice, I see you've now covered this case.
> On top of checking that VPs have been triggered we should also check that the corresponding Action has been written to storage to ensure that the VP will run the correct validation

Nice, I see you've also covered this case.

Reply:
> For a single transfer I think we would need do something like this:
> 
>     1. If a VP of an address who must authorize the tx is not going to be triggered from the changed keys (yes, it is only the leading part) then the tx has to `insert_verifier`
> 
>     2. The MASP VP should require an `Action` to be written so a tx is not allowed to skip it (i.e. actions must not be empty)
> 
> 
> For these to work together the MASP VP should then ensure that when the action is `MaspSigner` then the address in this action is contained in the `verifiers` set.

At least in the current PR, this is all covered right? The current conditional forces the presence of the signer in the verifiers list. The next conditional forces the presence of the corresponding `MaspSigner` `Action`.

> However, with the scenario with multiple transfers @murisi described I think this may not be sufficient.

The `signer` variable in this loop takes on as its value each of the `Address`es whose balances are debited (even if their net change is zero), so if this verifier check and the following `MaspSigner` check is correct then this VP should prevent Christel from committing thievery (since the `Tx` would be forced to trigger the signature checks elsewhere).